tests/federation_state_resolution_test.go

```diff
@@ -0,0 +1,179 @@
+package tests
+
+import (
+	"encoding/json"
+	"io"
+	"testing"
+	"time"
+
+	"github.com/matrix-org/complement"
+	"github.com/matrix-org/complement/b"
+	"github.com/matrix-org/complement/client"
+	"github.com/matrix-org/complement/ct"
+	"github.com/matrix-org/complement/federation"
+	"github.com/matrix-org/complement/helpers"
+	"github.com/matrix-org/gomatrixserverlib/spec"
+	"github.com/tidwall/gjson"
+)
+
+// Tests state resolution (v2) by forcing a fork with conflicted state, per:
+// https://github.com/matrix-org/matrix-spec/issues/1209
+// After initial room create and power levels change to enable name setting, we
+// have the following:
+//
+// alice sets name1
+//
+//	|
+//	|---------------------------.
+//	|                           |
+//	v                           v
+//
+// alice sets name2       bob kicks alice (this arrives late)
+//
+//	|                           |
+//	|     .---------------------'
+//	v     v
+//
+// what is the current state here
+func TestFederationStateResolution(t *testing.T) {
+	deployment := complement.Deploy(t, 1)
+	defer deployment.Destroy(t)
+	srv := federation.NewServer(t, deployment,
+		federation.HandleKeyRequests(),
+		federation.HandleMakeSendJoinRequests(),
+		federation.HandleTransactionRequests(nil, nil),
+	)
+	srv.UnexpectedRequestsAreErrors = false
+	cancel := srv.Listen()
+	defer cancel()
+
+	hsName := "hs1"
```

Comment:
I'd be tempted just to inline this.

The placement here makes it bit weird to figure out whether it pertains to the real homeserver from the deployment vs the engineered homeserver from Complement.